src/openms/source/DATASTRUCTURES/String.cpp

```diff
@@ -60,6 +60,11 @@ namespace OpenMS
   {
   }
 
+  String::String(const unsigned char* s) :
+    string(reinterpret_cast<const char*>(s))
```

Comment:
This sounds dangerous. At least the explicit keyword could save us from trouble.

Reply:
I'm happy to add the `explicit`, but none of the other String constructors is declared with it. Do you want it just for this one?

Reply:
I would love to have it for every single argument constructor - but requesting that sounds like asking for a lot ;)

Reply:
So I tried that - making every single-argument c'tor `explicit` - for String. First, for other string types it actually makes sense to allow implicit conversion - otherwise you need lots of overloads for the assignment operator, or you can't even assign e.g. a string literal. For numeric types, it would probably make sense to have `explicit` c'tors, but it leads to many test errors, so making the change now is impractical.

I've declared the new c'tor as `explicit`, though.

Reply:
this is definitely not something we should do routinely, in my mind `unsigned char` is not generally a string but raw binary data. Only if you are really sure that this is a string should you do something like this and use `reinterpret_cast`.  Therefore doing this by default does also seem dangerous to me, even with `explicit` in place...

However, I also learned that `char` is not specified to be signed or unsigned, but to me it seems clear that the intention of `char` and `signed char` would be to process ASCII data while `unsigned char` is raw data of unknown interpretation (a bit like Python string and byte arrays).

Reply:
> in my mind unsigned char is not generally a string but raw binary data

I don't think that's generally true. E. g. the SQLite C API returns `unsigned char*` for a text column, but `void*` for a BLOB, i.e. raw binary data.

Pragmatically, I think it's clearer to use the (explicit) String constructor than to litter the code with `reinterpret_cast` calls.

Reply:
> I don't think that's generally true. E. g. the SQLite C API returns `unsigned char*` for a text column, but `void*` for a BLOB, i.e. raw binary data.

the reason here being that this allows UTF8 support: https://stackoverflow.com/questions/319790/sqlite-c-c-api-unsigned-char (where UTF8 *is* a raw binary data that needs interpretation). Basically the conceptual problem here is that you get raw binary data and it makes sense to *explicitly* say how you want to interpret that raw data, in this case it is a choice to interpret that data as ASCII

Reply:
Fair point, but in my opinion an *explicit* String constructor also says that.